source/auth/auth.rst

```diff
@@ -1281,6 +1265,17 @@ mechanism_properties
         Drivers MUST allow the user to specify a callback of the form
         "onOIDCRefreshToken" (defined below), if the driver supports
         providing objects as mechanism property values.
+    ALLOWED_HOSTS
+        The list of allowed hostnames or ip-addresses (ignoring ports) for
+        MongoDB connections, allowing Unix shell-style wildcards. This is a
```

Comment:
We should specify what we mean by "Unix shell-style wildcards". Should the impl support `*` (yes), `?` (maybe), and `[]` (possibly)? What about `[a-d]`? How about `[[:upper:]]`? `[![:digit:]]`? I suspect that we want to support `*` and maybe `?`, but nothing else.

Furthermore when it comes to `*`, how greedy is the matching? If you specify:

```
*.example.com
```

which of the following will be successfully matched?

```
dev.example.com
dev.subdomain.example.com
evildomain.info?ignored=example.com
```

First one, yes. Second one, probably not, but maybe. Third one definitely not, but we should have some tests to verify.

Reply:
I based it on the semantics of `fnmatch` in [Python](https://docs.python.org/3/library/fnmatch.html#module-fnmatch), which closely matches the glob wiki [page](https://en.wikipedia.org/wiki/Glob_(programming)).  We could make it more explicit here:

<img width="402" alt="image" src="https://user-images.githubusercontent.com/2096628/230471958-246af3ca-c552-4433-853b-a24fcd0a7b52.png">

Reply:
NOTE: I'm not sure whether any language has easy support for this filtering. At this point, in c# impl I implement this logic manually. Maybe regex can be more portable since I would imagine it's more common functionality for all languages?

Reply:
Yeah, Spencer also initially mentioned that this could be a regex; the reason why I preferred not to do that is that it can be hard to get a regex “right” and not have it accidentally match more than intended.

My suggestion here would be to only allow `*` as a wildcard, with the semantics of `[a-zA-Z0-9:.]+` in regex language (i.e. match characters that are valid in domain names or IP addresses, ignoring things like IDNA).

Reply:
I agree with @addaleax on all points

Reply:
On second thought, @DmitryLukyanov and think we should go with regex: it is easier and less error-prone than implementing our own custom logic, and easier for the user to reason about and test using standard tools.

Reply:
It might help if we considered some examples (as in @JamesKovacs 's original comment)

     "www.mongodbgov.net",

If we use regex, wouldn't this match `wwwxmongodbgov.net`? This seems too easy to get wrong.

Reply:
From another side, it's easy to check whether particular host matches to provided expression :)

Reply:
If we do go with the regex approach (and I am not entirely convinced that that’s a good idea, but also practically speaking want to see this PR move forward of course), are we convinced that all drivers have access to regex libraries with reasonably similar semantics? Or are we okay with the same ALLOWED_HOSTS string resulting in slightly different validation rules with different drivers?

Reply:
What if we only allowed a leading greedy asterisk on the parsed host name, since that matches the concept of a subdomain.  In James' example, the first two would match and the last would not, because the parsed host name would be `evildomain.info`:

<img width="409" alt="image" src="https://user-images.githubusercontent.com/2096628/231591846-690ca81b-f25f-46ae-aa5e-de1802cd39a9.png">